src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp

```diff
@@ -321,40 +301,32 @@ inline T Atomic::PlatformCmpxchg<4>::operator()(T exchange_value,
   // specified otherwise (see atomic.hpp).
 
   T old_value;
-  const uint64_t zero = 0;
+  int is_successful;
 
   pre_membar(order);
-
-// FIXME_RISCV
-#if 0
   __asm__ __volatile__ (
     /* simple guard */
-    "   lwz     %[old_value], 0(%[dest])                \n"
-    "   cmpw    %[compare_value], %[old_value]          \n"
-    "   bne-    2f                                      \n"
+    "   lw     %[old_value], (%[dest])                          \n"
+    "   bne    %[compare_value], %[old_value], 2f               \n"
     /* atomic loop */
-    "1:                                                 \n"
-    "   lwarx   %[old_value], %[dest], %[zero]          \n"
-    "   cmpw    %[compare_value], %[old_value]          \n"
-    "   bne-    2f                                      \n"
-    "   stwcx.  %[exchange_value], %[dest], %[zero]     \n"
-    "   bne-    1b                                      \n"
+    "1:                                                         \n"
+    "   lr.w    %[old_value], (%[dest])                         \n"
+    "   bne     %[compare_value], %[old_value], 2f              \n"
+    "   sc.w    %[is_successful], %[exchange_value], (%[dest] ) \n"
+    "   bnez    %[is_successful], 1b                            \n"
     /* exit */
-    "2:                                                 \n"
+    "2:                                                         \n"
     /* out */
-    : [old_value]       "=&r"   (old_value),
-                        "=m"    (*dest)
+    : [old_value]       "=&r"   (old_value)
     /* in */
-    : [dest]            "b"     (dest),
-      [zero]            "r"     (zero),
+    : [dest]            "r"     (dest),
       [compare_value]   "r"     (compare_value),
       [exchange_value]  "r"     (exchange_value),
-                        "m"     (*dest)
+      [is_successful]   "r"     (is_successful)
```

Comment:
Please fix this.

`is_successful` register is written, so it should be in `out` list (here it's in `in`). It should be also have `"=&r"` constraint: it's written out (`=`) and it should differs from `in` registers (`&`). 

Without `&`, gcc is allowed to allocate `is_successful` to be equal to `dest`, for example. So the code will be equal to 
```
1:
lr.w    %[old_value], (%t0) // correct on 1st, incorrect on 2nd+ iteration
bne     %[compare_value], %[old_value], 2f
sc.w    %t0, %[exchange_value], (%t0) // destroys address to load from/store to
bnez    %t0, 1b
```

Such issue will be hard to debug down to this function. So we better spend some time to reviewing/fixing than to  considerably longer debugging later.

Reply:
Fixed